Updates to the contributor admin dashboard

stevenjgodfrey opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes issues in the new version of contributor dashboard admin, B3, B5, B6, B7 and E1 refer to the document in the contributor experience team documents section.
2. This PR fixes the following issues on the new version of contributor dashboard admin: 
    B3 - If a user is a question coordinator, but not a translation coordinator, the number of question reviewers appears too far below the banner in the box.
   B5 - The second-last column in the users table does not make sense (it should say “Last Reviewed” but this is a table of Question Submitters).
   B6 - In the users table, if there is a large amount of data in the final filed, the box does not extend down around it.
   B7 - I opened the roles for a user and removed the question reviewer role. I saved the changes. 
 When I reopened the modal, both the question submitter and reviewer roles were unchecked, even though I had only removed the reviewer
   E1 - I have not selected any language, so what does this number depict? Translation reviewer and Question reviewer for which language and which topics or overall/total number.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [x] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)



## Proof that changes are correct


https://github.com/oppia/oppia/assets/123914818/1275d429-5feb-4246-aafd-a66ef07a0d94

stevenjgodfrey committed:
Changes to contributor-dashboard-admin to appear correctly, to reduce number of calls to fetchContributorAdminStats, to subscribe in the correct order

stevenjgodfrey committed:
Add a test to contributor-dashboard-admin unit tests to ensure coverage

stevenjgodfrey committed:
Merge remote-tracking branch 'upstream/develop' into cont-dash-admin-change
merge user_services

stevenjgodfrey committed:
increase coverage of contributor-admin-dashboard page tests

stevenjgodfrey committed:
increase coverage of contributor-admin-dashboard page tests further

stevenjgodfrey committed:
Merge remote-tracking branch 'upstream/develop' into cont-dash-admin-change
Merge upstream

stevenjgodfrey committed:
Merge remote-tracking branch 'upstream/develop' into cont-dash-admin-change
merge with develop

stevenjgodfrey committed:
fix e2e test error by reducing elements to 1, as unseen component removed. Removed call to classroom/math as it is no longer added by default.

stevenjgodfrey committed:
Merge branch 'develop' into cont-dash-admin-change

stevenjgodfrey committed:
remove browser debug

stevenjgodfrey committed:
acl decorator changed to open access for classroom id to name

stevenjgodfrey committed:
fixed classroom linting error

stevenjgodfrey committed:
fixed further linting error in classroom

stevenjgodfrey committed:
fix typescipt check error in contributor-dashboard-admin-stats-backend-api.service test

stevenjgodfrey committed:
fix typescript check failure in contributor-dashboard-admin-stats-backend service

stevenjgodfrey committed:
Merge remote-tracking branch 'upstream/develop' into cont-dash-admin-change

stevenjgodfrey committed:
fix code coverage failure

stevenjgodfrey committed:
removed unnecessary test, and export

stevenjgodfrey committed:
Merge remote-tracking branch 'upstream/develop' into cont-dash-admin-change
merge develop

stevenjgodfrey committed:
add coverage to contributor dashboard admin test

stevenjgodfrey committed:
Merge branch 'cont-dash-admin-change' of https://github.com/stevenjgodfrey/oppia into cont-dash-admin-change
merge in from develop

stevenjgodfrey commented:
@seanlip PTAL

stevenjgodfrey committed:
Merge branch 'develop' into cont-dash-admin-change

stevenjgodfrey committed:
Merge branch 'develop' into cont-dash-admin-change

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -30,6 +30,12 @@ import { CdAdminQuestionRoleEditorModal } from './question-role-editor-modal/cd-
 import { UsernameInputModal } from './username-input-modal/username-input-modal.component';
 import { BrowserDynamicTestingModule } from '@angular/platform-browser-dynamic/testing';
 import { UserInfo } from 'domain/user/user-info.model';
+import {ContributorAdminStatsTable } from './contributor-dashboard-tables/contributor-admin-stats-table.component';
```
Nit: There should be a whitespace char between `{` and `C` of `Contributor...` on line 33.

stevenjgodfrey replied:
space added

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
 
-    expect(() => {
+    it('should not update translation and question coordinator view' +
+      'if username is null', fakeAsync(() => {
```
What does a `null` username mean on a higher, conceptual level? Once that is figured out, please reword the "username is null" part of the test name.

stevenjgodfrey replied:
changed to user is not logged in

stevenjgodfrey replied:
Also change username is not null to user is logged in

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
+  });
 
-    component.toggleLanguageDropdown();
+  describe('null language', () => {
```
What does a `null` language mean on a conceptual level? When you have an answer, please use that answer to replace the describe text with.

Also, try starting the describe text with `'when {insert-condition-here}'`

stevenjgodfrey replied:
changed to when no languages are available

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
 
-    expect(component.languageDropdownShown).toBeTrue();
-  });
+    it('should find language en if no language is populated ' +
```
1) Replace `find language en` to `default to English language`.
2) What does `username is not null` mean on a conceptual level?

stevenjgodfrey replied:
Changed to default to English language and when user is not logged in

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
+  });
 
-    component.toggleActivityDropdown();
+  describe('Not null username', () => {
```
What does `Not null username` mean on a conceptual level? Please reword the describe text with the answer to my question.

Also begin the describe text with `'when ...'`.

stevenjgodfrey replied:
Changed to when user is logged in

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
 
-    expect(component.selectedLanguage.language).toBe('English');
-    expect(component.selectedLanguage.id).toBe('en');
+    it('should call fetchContributorAdminStats only once', fakeAsync(() => {
```
1) Replace `call fetchContributorAdminStats` to `fetch contributor admin stats`. We want to know how the component behaves, not what function calls are being made.
2) Please indicate when this scenario occurs. The `ngOnInit` in this test is a good place to start.

stevenjgodfrey replied:
improved comment to be more meaningful

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
-    component.ngOnInit();
-    tick();
-    fixture.detectChanges();
+    it('should create ContributorAdminStatsTable only once', fakeAsync(() => {
```
Please indicate when this scenario occurs in the test name.

stevenjgodfrey replied:
updated test name to be more meaningful

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
-    expect(component.selectedLanguage.language).toBe('العربية (Arabic)');
-    expect(component.selectedLanguage.id).toBe('ar');
-  }));
+    it('should initialize variables onInit', fakeAsync(() => {
```
Please state in the test name what those initialized variables are.

StephenYu2018 replied:
Reminder to leave a reply to this thread once @stevenjgodfrey has fixed this.

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
-    expect(component.selectedLanguage.id).toBe('ar');
-    expect(component.selectedTopicIds).toEqual(['1', '2']);
-  }));
+      expect(component.selectedLanguage.language).toBe('English');
```
These two assertions before `component.selectLanguage` are unnecessary and make the test harder to read. Try to follow the "Arrange, Act, Assert" pattern when writing these types of tests.

stevenjgodfrey replied:
Removed extra assertions

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
 
-    expect(component.selectedContributionType).toEqual('selection1');
-  });
+      component.selectLastActivity(7);
```
Maybe try this instead:
```ts
const activity = 7
...
component.selectLastActivity(activity);
expect(component.selectedLastActivity).toEqual(activity);
```
That way, if the data type or values of `activity` ever changes in the future, the developer who changes them can easily change the test data in the future. Also, the word `activity` provides explicit meaning compared to literals and numbers.

Try the same thing for the second activity in this test.

stevenjgodfrey replied:
Added variable

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
-    component.ngOnInit();
-    tick();
-    fixture.detectChanges();
+    it('should remove duplicates and apply topic filter', fakeAsync(() => {
```
Duplicates of what? Topics? Please clarify this in the test name.

Also, I'm not seeing any duplicate test data within this test.

stevenjgodfrey replied:
Test should really be 2 tests, so I've split it.

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
-    component.ngOnInit();
-    tick();
-    fixture.detectChanges();
+    it('should throw error when topic filter chooses null', fakeAsync(() => {
```
What does it mean for the topic filter to choose `null`? Does it mean that the app doesn't apply any filters based on topic names?

stevenjgodfrey replied:
I think this is trying to provide code coverage for the code below 

if (!matchingTopic && this.selectedTopicNames.length) {
          throw new Error(
            'Selected Topic Id doesn\'t match any valid topic.');
        }
        
This seems an unlikely scenario, but possible given more that one topic filter can be added at once. I've changed the description to better indicate this.

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
+      expect(component.selectedContributionType).toEqual('selection1');
+    });
+
+    it('should update translation coordinator view', fakeAsync(() => {
```
When should the component update translation coordinator view?

stevenjgodfrey replied:
updated the description

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
-        fakeClickAwayEvent,
-        'target',
-        {value: document.createElement('div')});
+    it('should update question coordinator view', fakeAsync(() => {
```
When should the component update question coordinator view?

stevenjgodfrey replied:
updated the description

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
+        expect(component.isTranslationCoordinator).toBeTrue();
+      }));
+
+    it('should correctly show and hide language dropdown when clicked away',
```
Remove 'correctly show and' and add 'opened' before 'language dropdown'

stevenjgodfrey replied:
updated the comment

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
+        expect(component.languageDropdownShown).toBe(false);
+      }));
+
+    it('should correctly show and hide activity dropdown when clicked away',
```
Remove 'correctly show and' and add 'opened' before 'activity dropdown'

stevenjgodfrey replied:
expectation changed

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
+        expect(component.activityDropdownShown).toBe(false);
+      }));
+
+    it('should show and hide activity dropdown when clicked away on review tab',
```
If this test doesn't assert anything about `component.activeTab`, then how is this test different from the previous? Also, are we trying to say when the user clicks away to the review tab? Or when the user is on the review tab and clicks away from the dropdown? Please clarify.

stevenjgodfrey replied:
removed

seanlip commented:
@chris7716 @StephenYu2018  Could you please do the initial review for this? Thanks.

StephenYu2018 commented:
@seanlip Yep. I already started.

StephenYu2018 commented on core/templates/domain/classroom/classroom-backend-api.service.spec.ts:
```diff
@@ -216,7 +216,7 @@ describe('Classroom backend API service', function() {
       successHandler, failHandler);
     let req = httpTestingController.expectOne(
-      '/classroom_admin_data_handler');
+      '/classroom_id_to_name_handler');
```
Please use the appropriate constant defined in `core/templates/domain/classroom/classroom-domain.constants.ts` as opposed to using string literals. That way if the constant's value needs to change, the developer changing said constant only needs to look in one place to change it (in this case, it would be the mentioned constants file).

stevenjgodfrey replied:
Added

StephenYu2018 commented on core/templates/domain/classroom/classroom-backend-api.service.spec.ts:
```diff
@@ -244,7 +244,7 @@ describe('Classroom backend API service', function() {
         successHandler, failHandler);
       let req = httpTestingController.expectOne(
-        '/classroom_admin_data_handler');
+        '/classroom_id_to_name_handler');
```
Please use the appropriate constant defined in `core/templates/domain/classroom/classroom-domain.constants.ts` as opposed to using string literals. That way if the constant's value needs to change, the developer changing said constant only needs to look in one place to change it (in this case, it would be the mentioned constants file).

stevenjgodfrey replied:
renamed

StephenYu2018 commented on core/templates/domain/classroom/classroom-backend-api.service.ts:
```diff
@@ -223,7 +223,7 @@ export class ClassroomBackendApiService {
     return new Promise((resolve, reject) => {
       this.http.get<ClassroomIdToClassroomNameBackendDict>(
-        '/classroom_admin_data_handler').toPromise().then(response => {
+        '/classroom_id_to_name_handler').toPromise().then(response => {
```
Please use the appropriate constant defined in `core/templates/domain/classroom/classroom-domain.constants.ts` as opposed to using string literals. That way if the constant's value needs to change, the developer changing said constant only needs to look in one place to change it (in this case, it would be the mentioned constants file).

stevenjgodfrey replied:
used constant

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.html:
```diff
@@ -107,17 +107,17 @@ <h2 class="stat-value-with-rating e2e-test-oppia-total-plays">
           <div class="topic-selector-container">
             <oppia-topic-filter
-              *ngIf="(activeTab === TAB_NAME_TRANSLATION_SUBMITTER) || (activeTab === TAB_NAME_QUESTION_SUBMITTER)"
+              *ngIf="topicsFetched==='true' && (activeTab === TAB_NAME_TRANSLATION_SUBMITTER) || (activeTab === TAB_NAME_QUESTION_SUBMITTER)"
```
Nit: Needs 1 whitespace character between `===`

stevenjgodfrey replied:
added

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -88,6 +89,7 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
   selectedTopicNames: string[] = [];
   languageChoices: LanguageChoice[] = [];
   topics: TopicChoice[] = [];
+  topicsFetched: string = 'false';
```
By what I'm observing, `topicsFetched` seems to be used like a boolean value. And if that's intentional, then it's better to use actual boolean values as opposed to strings. Also, a more readable name for the boolean variable would be `topicsAreFetched` as derived from [this coding style guideline](https://github.com/oppia/oppia/wiki/Coding-style-guide#javascript).

stevenjgodfrey replied:
I've renamed the variable, but I can't use a boolean value, as in the component.html it is found to have a value of TRUE (because it exists), even if the value defined in componenet.ts FALSE. Hope this makes sense, feel free to get in touch to discuss if it doesn't :)

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.html:
```diff
@@ -190,7 +190,7 @@ <h2 class="stat-value-with-rating e2e-test-oppia-total-plays">
         </div>
       </div>
-      <contributor-admin-stats-table [activeTab]="activeTab" [filter]="filter"></contributor-admin-stats-table>
+      <contributor-admin-stats-table  [inputs]="{ activeTab: activeTab, filter: filter }"></contributor-admin-stats-table>
```
Nit: One unnecessary whitespace character before `[inputs]`

stevenjgodfrey replied:
removed

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-dashboard-tables/contributor-admin-stats-table.component.html:
```diff
@@ -485,13 +485,11 @@ <h2>{{ noDataMessage }}</h2>
-    height: 16px;
     justify-content: flex-start;
-  }
+    }
```
Why the extra indent?

stevenjgodfrey replied:
removed

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-dashboard-tables/contributor-admin-stats-table.component.html:
```diff
@@ -714,5 +711,10 @@ <h2>{{ noDataMessage }}</h2>
     }
-  }
+
+    .attribute-key {
+      min-width: 50%;
+    }
```
I think these also have an extra 2 space indent.

stevenjgodfrey replied:
removed

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
-    expect(modalSpy).toHaveBeenCalledWith(CdAdminQuestionRoleEditorModal);
-    expect(addRightsSpy).toHaveBeenCalled();
-  }));
+    it('should open question role editor modal and return changed value' +
```
I'm not seeing anything about translation reviewers involved in this test. Is the test name incorrect for mentioning translation reviewers, or is the test using the wrong values?

stevenjgodfrey replied:
changed expectation of test

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
-    expect(addRightsSpy).toHaveBeenCalled();
-  }));
+    it('should open question role editor modal and return changed value of' +
+      ' translation reviewer', fakeAsync(() => {
```
I'm not seeing anything about translation reviewers involved in this test. Is the test name incorrect for mentioning translation reviewers, or is the test itself using the wrong test data?

stevenjgodfrey replied:
changed expectation of test

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -117,76 +119,93 @@ export class ContributorAdminDashboardPageComponent implements OnInit {
+            this.contributorDashboardAdminStatsBackendApiService
+              .fetchTopicChoices().then(
+                response => {
+                  this.topics = response.flat();
```
Is it possible for `response.flat()` to be in the line below line 181? I feel like the extra state mutation isn't necessary here.

stevenjgodfrey replied:
moved response.flat to the line below

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/services/contributor-dashboard-admin-stats-backend-api.service.spec.ts:
```diff
@@ -404,25 +472,61 @@ describe('Contribution Admin dasboard stats service', () => {
+  it('should return data for all classrooms', fakeAsync(
     () => {
-      const url = '/classroom_data_handler/math';
+      spyOn(cdasbas, 'fetchTopicChoices').and.callThrough();
```
It seems like `fetchTopicChoices` is the action under test, so I don't think this spy is necessary.

stevenjgodfrey replied:
removed

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/services/contributor-dashboard-admin-stats-backend-api.service.spec.ts:
```diff
@@ -404,25 +472,61 @@ describe('Contribution Admin dasboard stats service', () => {
-  it('should return topics data in math classroom', fakeAsync(
+  it('should return classroom data for a classroom id', fakeAsync(
+    () => {
+      spyOn(cdasbas, 'fetchTopics').and.callThrough();
```
It seems like `fetchTopics` is the action under test, so I don't think this spy is necessary.

stevenjgodfrey replied:
removed

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/services/contributor-dashboard-admin-stats-backend-api.service.spec.ts:
```diff
@@ -404,25 +472,61 @@ describe('Contribution Admin dasboard stats service', () => {
+      spyOn(crbas, 'fetchClassroomDataAsync')
+        .and.returnValue(Promise.resolve(sampleClassroomDataObject));
+
+      cdasbas.fetchTopics('0').then(
+        successHandler, failHandler
+      );
+      flushMicrotasks();
+
+      expect(crbas.getClassroomDataAsync).
+        toHaveBeenCalledWith('0');
```
Replace the literal `'0'` with a variable/constant named after what the literal is supposed to represent in the context of this test. Since the interface of this service says that `fetchTopics` requires a `classroomId`, the name `classroomId` should suffice.

stevenjgodfrey replied:
changed to classroomId

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/services/contributor-dashboard-admin-stats-backend-api.service.spec.ts:
```diff
@@ -404,25 +472,61 @@ describe('Contribution Admin dasboard stats service', () => {
-      const url = '/classroom_data_handler/math';
+      spyOn(cdasbas, 'fetchTopicChoices').and.callThrough();
+      spyOn(crbas, 'getAllClassroomIdToClassroomNameDictAsync')
+        .and.returnValue(Promise.resolve({mathclassroomId: 'math'}));
```
It should be spelled `'mathClassroomId'` to keep consistent with `camelCase`.

stevenjgodfrey replied:
changed to mathClassroomId

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -106,438 +135,531 @@ describe('Contributor dashboard Admin page', () => {
+      }));
+
+    it('should open question role editor modal when on question' +
+      ' submitter/reviewer tabs', fakeAsync(() => {
```
Remove `/reviewer` and the `s` from `tabs` in the test name. There's already another test covering this tab type.

stevenjgodfrey replied:
removed

StephenYu2018 reviewed (changes requested):
I've left several nits and concerns. Please take a look.

stevenjgodfrey committed:
Merge remote-tracking branch 'upstream/develop' into cont-dash-admin-change
merge from develop

stevenjgodfrey committed:
Improve test files to make language used more meaningful

stevenjgodfrey committed:
Merge remote-tracking branch 'upstream/develop' into cont-dash-admin-change
merge with develop

stevenjgodfrey committed:
tidied up use of domain constants and expectations of tests

stevenjgodfrey committed:
Merge branch 'cont-dash-admin-change' of https://github.com/stevenjgodfrey/oppia into cont-dash-admin-change
pull from remote

stevenjgodfrey commented:
@StephenYu2018 I've made the requested changes, PTAL

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.html:
```diff
@@ -107,17 +107,17 @@ <h2 class="stat-value-with-rating e2e-test-oppia-total-plays">
         </div>
       </div>
-      <contributor-admin-stats-table [activeTab]="activeTab" [filter]="filter"></contributor-admin-stats-table>
+      <contributor-admin-stats-table [inputs]="{ activeTab: activeTab, filter: filter }"></contributor-admin-stats-table>
```
I'm pretty sure that `{ activeTab, filter }` will work due to JS Object syntax. I'd prefer writing it that way, since there's no code duplication.

Same can be said with line 193 of this file.

stevenjgodfrey replied:
Turns out this notation { activeTab: activeTab, filter: filter } is required, otherwise the tests fail

StephenYu2018 replied:
Alright. Keep it that way then.

StephenYu2018 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-dashboard-tables/contributor-admin-stats-table.component.html:
```diff
@@ -714,5 +711,10 @@ <h2>{{ noDataMessage }}</h2>
+
+    .attribute-key {
+      min-width: 50%;
   }
+
+}
```
Lines 717 and 719 still need 2 spaces of indentation.

stevenjgodfrey replied:
added

StephenYu2018 reviewed (changes requested):
A few more things before it's good to go.

stevenjgodfrey committed:
Merge remote-tracking branch 'upstream/develop' into cont-dash-admin-change
merge with develop

stevenjgodfrey committed:
fixed indentation issue

stevenjgodfrey commented:
@StephenYu2018 PTAL. I've addressed the requested changes, Thanks for the review!

StephenYu2018 reviewed (approved):
LGTM!

seanlip commented on core/tests/webdriverio_desktop/contributorAdminDashboard.js:
```diff
@@ -286,12 +286,12 @@ describe('Contributor Admin Dashboard', function() {
     await contributorDashboardAdminPage.navigateToQuestionSubmitterTab();
     await contributorDashboardAdminPage.waitForLoadingMessageToDisappear();
-    await contributorDashboardAdminPage.expectStatsElementCountToBe(2);
+    await contributorDashboardAdminPage.expectStatsElementCountToBe(1);
```
Just to check -- why did this change?

stevenjgodfrey replied:
This change is because the stats component was being drawn twice on the page, once for mobile and once for desktop. I changed it so that it  is being drawn once, either mobile or desktop, I don't think it breaks anything, I tried asking Kshitij for some input but haven't heard back yet. The change is one of a number I made to reduce the number of times that the stats component was being drawn and the associated api call being made  ( I amalgamated the variables passed in to the stats component into 1 variable, added the topicsAreFetched variable, made the filter only change when something on it had changed). It was being drawn 8 times, its down to 2 now, I don't think it can go any lower due to the angular js component.

seanlip replied:
Thanks -- LGTM. Just one concern here ... it sounds like this is set up to count hidden elements that the user doesn't see (since if they are on mobile they won't see desktop, and vice versa).

That seems wrong, shouldn't the test be structured to only detect stuff that is actually visible to the user?

stevenjgodfrey replied:
I've already added this *ngIf="checkMobileView() in the contributor-admin-dashboard.component.html page, so that only the mobile view is shown, and hence the hidden element is no longer present. There was already a check on *ngIf="!checkMobileView() for the desktop view. 

I was a little concerned because the e2e test ( the one you have highlighted ) specifically checked to see that 2 components were present, even though only 1 is visible. But I did some manual testing around this and i can't see how anything is broken.

seanlip replied:
Hm, ok, that's weird, I wonder whether the test is actually detecting the right thing?

Though if you feel it's all good, no concerns about merging. Thanks!

seanlip commented on core/domain/user_services.py:
```diff
@@ -668,7 +668,8 @@ def _update_user_contribution_rights(
     """
-    if user_contribution_rights.can_review_at_least_one_item():
+    if user_contribution_rights.can_review_at_least_one_item() or (
+       user_contribution_rights.can_submit_at_least_one_item()):
```
I think the docstring needs to be updated?

stevenjgodfrey replied:
Added

seanlip commented on core/domain/user_domain.py:
```diff
@@ -1204,6 +1204,14 @@ def can_review_at_least_one_item(self) -> bool:
             self.can_review_voiceover_for_language_codes or
             self.can_review_questions)
 
+    def can_submit_at_least_one_item(self) -> bool:
```
Any unit tests for this?

stevenjgodfrey replied:
I've added a few tests for userContributionRights

seanlip reviewed (changes requested):
Thanks @stevenjgodfrey! Sorry for the late reply. Reviewed codeowner files.

kevintab95 reviewed (approved):
LGTM for codeowner files. Thanks!

stevenjgodfrey committed:
merge with develop, includes manuial merge main.py and classroom.py

stevenjgodfrey committed:
finish manual merge of main and classroom

stevenjgodfrey committed:
added tests for userContributionRights, and changed description

stevenjgodfrey committed:
removed extra blank lines from classroom.py

stevenjgodfrey commented:
> Thanks @stevenjgodfrey! Sorry for the late reply. Reviewed codeowner files.

@seanlip PTAL. 
Thanks for the review, I've addressed the comments.

stevenjgodfrey committed:
Merge branch 'develop' into cont-dash-admin-change

seanlip reviewed (approved):
Thanks @stevenjgodfrey! Looks good -- just one comment, please address before merging. Thanks!

stevenjgodfrey commented:
@seanlip PTAL

I've replied to your comment, hopefully it addresses your concerns.
